File: packages/mobile-core/src/queue.ts
import { type Identity, defaultIdFactory } from "./identity";
import type { Batch, Clock, DeviceInfo, EventInput, Transport } from "./types";

/** Same cadence as the web snippet. */
export const FLUSH_SIZE = 10;
export const FLUSH_INTERVAL_MS = 5000;
/** /v1/events caps a batch at 50 events — split larger buffers. */
const MAX_BATCH = 50;

/**
 * Buffers events and flushes on size / interval / app-background. The ingest
 * response may carry a pending Live Assist — surfaced via the response handler.
 */
export class EventQueue {
  private buffer: EventInput[] = [];
  private timer: ReturnType<typeof setTimeout> | null = null;
  private onResponse?: (data: unknown) => void;
  private inflight: Promise<void> = Promise.resolve();

  constructor(
    private readonly key: string,
    private readonly eventsUrl: string,
    private readonly identity: Identity,
    private readonly device: DeviceInfo,
    private readonly transport: Transport,
    private readonly now: Clock,
  ) {}

  setResponseHandler(fn: (data: unknown) => void): void {
    this.onResponse = fn;
  }

  enqueue(event: EventInput): void {
    const identified = { ...event, eventId: event.eventId ?? defaultIdFactory("evt") };
    this.identity.touchSession();
    this.buffer.push(identified);
    if (this.buffer.length >= FLUSH_SIZE) {
      void this.flush();
    } else if (!this.timer) {
      this.timer = setTimeout(() => void this.flush(), FLUSH_INTERVAL_MS);
      // Don't hold a Node test process open (no-op on RN/Hermes).
      (this.timer as { unref?: () => void }).unref?.();
    }
  }

  /** Send everything buffered. Serialized so batches arrive in order. */
  flush(): Promise<void> {
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
    if (this.buffer.length === 0) return this.inflight;
    const events = this.buffer.splice(0, this.buffer.length);
    this.inflight = this.inflight.then(async () => {
      for (let i = 0; i < events.length; i += MAX_BATCH) {
        const batch: Batch = {
          key: this.key,
          anonId: this.identity.getAnonId(),
          userId: this.identity.getUserId(),
          sessionId: this.identity.currentSession(),
          sentAt: this.now(),
          device: this.device,
          events: events.slice(i, i + MAX_BATCH),
        };
        try {
          const data = await this.transport.post(this.eventsUrl, batch);
          if (data !== undefined) this.onResponse?.(data);
        } catch {
          /* fail-silent: behavioral telemetry must never break the host app */
        }
      }
    });
    return this.inflight;
  }

  /** Test helper. */
  get size(): number {
    return this.buffer.length;
  }
}

File: sdks/ios/Sources/TrackiSDK/Queue.swift
//
//  Queue.swift
//  Buffers events and flushes on size / interval / app-background. The ingest
//  response may carry a pending Live Assist — surfaced via the response handler.
//

import Foundation

/// Same cadence as the web snippet.
public let flushSize = 10
public let flushIntervalMs = 5000
/// /v1/events caps a batch at 50 events — split larger buffers.
let maxBatch = 50

public final class EventQueue: @unchecked Sendable {
    private let key: String
    private let eventsUrl: String
    private let identity: Identity
    private let device: DeviceInfo
    private let transport: Transport
    private let now: Clock

    private var buffer: [EventInput] = []
    private var timer: DispatchSourceTimer?
    private var onResponse: ((Any?) -> Void)?
    /// Serializes flushes so batches arrive in order (mirrors the TS inflight chain).
    private var inflight: Task<Void, Never> = Task {}
    private let lock = NSLock()

    public init(
        key: String,
        eventsUrl: String,
        identity: Identity,
        device: DeviceInfo,
        transport: Transport,
        now: @escaping Clock
    ) {
        self.key = key
        self.eventsUrl = eventsUrl
        self.identity = identity
        self.device = device
        self.transport = transport
        self.now = now
    }

    public func setResponseHandler(_ fn: @escaping (Any?) -> Void) {
        onResponse = fn
    }

    public func enqueue(_ event: EventInput) {
        identity.touchSession()
        lock.lock()
        buffer.append(event)
        let count = buffer.count
        lock.unlock()
        if count >= flushSize {
            _ = flushTask()
        } else if timer == nil {
            armTimer()
        }
    }

    private func armTimer() {
        let t = DispatchSource.makeTimerSource(queue: .global())
        t.schedule(deadline: .now() + .milliseconds(flushIntervalMs))
        t.setEventHandler { [weak self] in _ = self?.flushTask() }
        timer = t
        t.resume()
    }

    private func cancelTimer() {
        timer?.cancel()
        timer = nil
    }

    /// Send everything buffered. Serialized so batches arrive in order.
    @discardableResult
    public func flushTask() -> Task<Void, Never> {
        cancelTimer()
        lock.lock()
        if buffer.isEmpty {
            lock.unlock()
            return inflight
        }
        let events = buffer
        buffer.removeAll(keepingCapacity: true)
        lock.unlock()

        let prev = inflight
        let next = Task { [self] in
            await prev.value
            var i = 0
            while i < events.count {
                let end = min(i + maxBatch, events.count)
                let batch = Batch(
                    key: key,
                    anonId: identity.getAnonId(),
                    userId: identity.getUserId(),
                    sessionId: identity.currentSession(),
                    sentAt: now(),
                    device: device,
                    events: Array(events[i..<end])
                )
                let data = await transport.post(eventsUrl, batch.toJSON())
                if data != nil { onResponse?(data) }
                // fail-silent: transport already swallows errors and returns nil.
                i += maxBatch
            }
        }
        inflight = next
        return next
    }

    /// Await-friendly flush mirroring the reference's `flush(): Promise<void>`.
    public func flush() async {
        await flushTask().value
    }

    /// Test helper.
    public var size: Int {
        lock.lock(); defer { lock.unlock() }
        return buffer.count
    }
}
